### PR TITLE
Fix zlib EOB condition

### PR DESCRIPTION
**Pull request by LemonBoy**
CC @leroycep, let me know if it works for you, thanks for discovering this (not-so) edge case I didn't take into account.

We lose a wee bit of speed but such is life :(

### CODE REVIEW

File: lib/std/compress/deflate.zig [hunk 1]
```diff
@@ -481,32 +483,52 @@ pub fn InflateStream(comptime ReaderType: type) type {
...
+                _ = try self.peekBits(code_len);
+                // Small optimization win, use as many bits as possible in the
+                // table lookup.
+                prefix = self.bits & ((1 << PREFIX_LUT_BITS) - 1);
+
+                const lut_entry = &h.prefix_lut[prefix];
+                // The code is longer than PREFIX_LUT_BITS!
+                if (lut_entry.len == 0)
+                    break;
+                // If the code lenght doesn't increase we found a match.
```
Comment (dbandstra) [inline lib/std/compress/deflate.zig:503]: Pointing out this "lenght" only because I saw you had fixed it in another place 😮

**Comment by leroycep** [comment]
It's working now, awesome! BTW, here's a testcase I put in `compress/zlib` to test it locally:

```zig
test "EndOfStream edge case" {
    const DATA = "08d763f8cfc0c000c1ffff43300303c3ffffff0183950bf5";
    var data: [DATA.len / 2]u8 = undefined;
    try std.fmt.hexToBytes(data[0..], DATA[0..]);
    
    var in_stream = io.fixedBufferStream(data[0..]);

    var zlib_stream = try zlibStream(testing.allocator, in_stream.reader());
    defer zlib_stream.deinit();

    // Read and decompress the whole file
    const buf = try zlib_stream.reader().readAllAlloc(testing.allocator, std.math.maxInt(usize));
    defer testing.allocator.free(buf);

    assertEqual("00ff000000ff000000ff00ffffff00ffffff00000000ffffff", buf);
}
```

I could probably convert it to use `testReader` if that would be preferred.

**Comment by LemonBoy** [comment]
> BTW, here's a testcase I put in compress/zlib to test it locally:

I've already added your test case to `zlib.zig` :wink:

File: lib/std/compress/deflate.zig (hunk 1 above)
Comment (LemonBoy) [reply lib/std/compress/deflate.zig:503]: After a decade or so I keep on misspelling that damn word :(

File: lib/std/compress/deflate.zig [hunk 2]
```diff
@@ -481,32 +483,52 @@ pub fn InflateStream(comptime ReaderType: type) type {
...
-                return h.prefix_lut[prefix];
+            // Using u32 instead of u16 to reduce the number of casts needed.
+            var prefix: u32 = 0;
+
+            // Fast path, read some bits and hope they're the prefix of some code.
+            // We can't read PREFIX_LUT_BITS as we don't want to read past the
+            // deflate stream end, use an incremental approach instead.
+            var code_len = h.min_code_len;
+            while (true) {
+                _ = try self.peekBits(code_len);
```
Comment (leroycep) [inline lib/std/compress/deflate.zig:494]: Could this be made faster by instead using `bits_left` and then clamping it between `min_code_len` and `PREFIX_LUT_BITS`?

Comment (LemonBoy) [reply lib/std/compress/deflate.zig:494]: Well it's already doing something like that if I understand your idea, the upper bound is not checked as the extra bits are masked away and the bogus `peekBits` is only used to make sure we have enough bits.

This part of the decoding process is extremely hot so please experiment with it, we may squeeze some other juice out of this without bigger changes (such as introducing secondary/tertiary lookup tables for longer codes).

Comment (leroycep) [reply lib/std/compress/deflate.zig:494]: Is there a way to test the performance of the implementation already set up?

Comment (LemonBoy) [reply lib/std/compress/deflate.zig:494]: I'm using this simple snippet:

```zig
const std = @import("std");
const io = std.io;
const mem = std.mem;
const math = std.math;
const zlib = std.compress.zlib;

pub fn main() !void {
    const DATA = @embedFile("/tmp/big.z");

    var fbs = io.fixedBufferStream(DATA);
    var buf: [16384]u8 = undefined;

    var zlib_stream = try zlib.zlibStream(std.heap.page_allocator, fbs.reader());
    defer zlib_stream.deinit();

    while (true) {
        const r = try zlib_stream.read(&buf);
        if (r == 0) break;
    }
}
```

Where `z.big` is obtained with `head -c10000000 /dev/urandom | xxd | zpipe > /tmp/big.z` and [this](https://github.com/madler/zlib/blob/master/examples/zpipe.c) is zpipe.

**Pull request merged**